src/pages/achievement/control/subcomponents/AchievementControlPanel.tsx

```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+
+import AchievementInferencer from '../../dashboard/subcomponents/utils/AchievementInferencer';
+import TaskAdder from './controlPanelTools/controlPanelUtils/TaskAdder';
+import TaskUploader from './controlPanelTools/controlPanelUtils/TaskUploader';
+import EditableAchievementTask from './controlPanelTools/EditableAchievementTask';
+
+type AchievementControlPanelProps = {
+  inferencer: AchievementInferencer;
+  updateAchievements: any;
+  forceRender: any;
+  isDisabled: boolean;
+
+  pendingUpload: any;
+  setPendingUpload: any;
+  saveAchievementsToFrontEnd: any;
+};
+
+function AchievementControlPanel(props: AchievementControlPanelProps) {
+  const {
+    inferencer,
+    updateAchievements,
+    forceRender,
+    isDisabled,
+    pendingUpload,
+    setPendingUpload,
+    saveAchievementsToFrontEnd
+  } = props;
+
+  const handleSaveChanges = () => {
+    setPendingUpload(true);
+    saveAchievementsToFrontEnd(inferencer.getAchievements());
+    forceRender();
+  };
+
+  const handleUploadChanges = () => {
```

Comment:
May be another place to use React.useCallback/React.useMemo. But make sure first that it makes a difference to the number of renders of TaskUploader/TaskAdder.